### PR TITLE
Fix pruning

### PR DESCRIPTION
**Pull request by varunagrawal**
Holding PR for getting the decision tree pruning to work.

### CODE REVIEW

**Comment by varunagrawal** [comment]
@dellaert @ProfFan please review this PR. It should be easier to do so since the intermediate changes (like comments, extraneous prints etc) are all removed.

**Comment by dellaert** [comment]
> @dellaert @ProfFan please review this PR. It should be easier to do so since the intermediate changes (like comments, extraneous prints etc) are all removed.

I will once you remove changes in discrete. That has been my comment over and over again and every time I get back to the review it's still there.

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 1]
```diff
@@ -72,73 +72,146 @@ void IncrementalHybrid::update(GaussianHybridFactorGraph graph,
...
   remainingFactorGraph_ = *result.second;
 
-  // Add the partial bayes net to the posterior bayes net.
-  hybridBayesNet_.push_back<HybridBayesNet>(*bayesNetFragment);
+  gttoc_(Elimination);
 
   // Prune
   if (maxNrLeaves) {
-    const auto N = *maxNrLeaves;
+    DecisionTreeFactor::shared_ptr discreteFactor = prune(*maxNrLeaves);
```
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: Explain what is happening in `prune`. Also, make functional.

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 2]
```diff
@@ -72,73 +72,146 @@ void IncrementalHybrid::update(GaussianHybridFactorGraph graph,
...
+
+    // If valid pruned discrete factor, then propagate to gaussian mixtures
```
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: Explain when does this happen better.

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 3]
```diff
@@ -72,73 +72,146 @@ void IncrementalHybrid::update(GaussianHybridFactorGraph graph,
...
+    if (discreteFactor) {
+      HybridBayesNet::shared_ptr prunedBayesNetFragment =
```
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: // pruneBayesNet sets the leaves with 0 in `discreteFactor` to nullptr in all the conditionals with the same keys in `bayesNetFragment`.

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp [hunk 4]
```diff
@@ -313,14 +355,234 @@ TEST_UNSAFE(DCGaussianElimination, Incremental_approximate) {
...
+  IncrementalHybrid inc;
+
+  // Regular ordering going up the chain.
+  Ordering ordering;
+  ordering += W(0);
+  ordering += Z(0);
+  ordering += Y(0);
+  ordering += X(0);
+
+  // Update without pruning
```
Comment (varunagrawal) [inline gtsam/hybrid/tests/testIncrementalHybrid.cpp:422]: // The result is a HybridBayesNet with no discrete variables (equivalent to a GaussianBayesNet).

Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:422]: Factorization is
`P(... | measurements) = P(W0|Z0) P(Z0|Y0) P(Y0|X0) P(X0)`

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp [hunk 5]
```diff
@@ -313,14 +355,234 @@ TEST_UNSAFE(DCGaussianElimination, Incremental_approximate) {
...
+  ordering += W(1);
+  ordering += Z(1);
+  ordering += Y(1);
+  ordering += X(1);
+
+  gfg = fg.linearize(initial);
+  fg = NonlinearHybridFactorGraph();
+
+  // Update without pruning
+  inc.update(gfg, ordering);
```
Comment (varunagrawal) [inline gtsam/hybrid/tests/testIncrementalHybrid.cpp:484]: // The result is a HybridBayesNet with 1 discrete variable M(1).
// `P(... | measurements) = P(W0|Z0, W1, M1) P(Z0|Y0, W1, M1) P(Y0|X0, W1, M1) P(X0 | X1, W1, M1) P(W1|Z1, X1, M1)P(Z1|Y1, X1, M1)P(Y1 | X1, M1)P(X1 | M1)P(M1)`

Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:484]: TODO(Frank): Fix printing of `GaussianConditional` to show the conditioned variables.

Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:484]: What is the MHS tree here?

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp [hunk 6]
```diff
@@ -313,14 +355,234 @@ TEST_UNSAFE(DCGaussianElimination, Incremental_approximate) {
...
+  ordering += W(2);
+  ordering += Z(2);
+  ordering += Y(2);
+  ordering += X(2);
+
+  gfg = fg.linearize(initial);
+  fg = NonlinearHybridFactorGraph();
+
+  // Now we prune!
+  inc.update(gfg, ordering, 2);
```
Comment (varunagrawal) [inline gtsam/hybrid/tests/testIncrementalHybrid.cpp:535]: What is the factorization?

// `P(... | measurements) = P(W0|Z0, W1, M1) P(Z0|Y0, W1, M1) P(Y0|X0, W1, M1) P(X0 | X1, W1, M1) P(W1|W2, Z1, X1, M1, M2) P(Z1| W2, Y1, X1, M1, M2)P(Y1 | W2, X1, M1, M2)P(X1 | W2, X2, M1, M2) P(W2|Z2, X2, M1, M2) P(Z2|Y2, X2, M1, M2)P(Y2 | X2, M1, M2)P(X2 | M1, M2)  P(M1, M2)`

File: gtsam/hybrid/IncrementalHybrid.cpp [hunk 7]
```diff
@@ -72,73 +72,146 @@ void IncrementalHybrid::update(GaussianHybridFactorGraph graph,
...
-        discreteFactor->enumerate();
+  return discreteFactor;
+}
 
-    // Loop over all assignments and create a vector of GaussianConditionals
-    std::vector<GaussianFactor::shared_ptr> prunedConditionals;
-    for (auto &&av : assignments) {
-      const DiscreteValues &assignment = av.first;
-      const double value = av.second;
+HybridBayesNet::shared_ptr IncrementalHybrid::pruneBayesNet(
```
Comment (varunagrawal) [inline gtsam/hybrid/IncrementalHybrid.cpp]: TODO(Frank) This should be a const method on HybridBayesNet.

File: gtsam/hybrid/tests/testIncrementalHybrid.cpp (hunk 6 above)
Comment (varunagrawal) [reply gtsam/hybrid/tests/testIncrementalHybrid.cpp:535]: The MHS at this point should be a 3 level tree on (0, 1, 2).
0 has 2 choices, 1 has 4 choices and 2 has 4 choices.

**Review by dellaert — APPROVED** [review]
Very cool!

**Comment by dellaert** [comment]
But please merge develop into both branches before merging and see if there are conflicts?

**Review by dellaert — CHANGES_REQUESTED** [review]
Same issue with this PR: many unrelated changes, seems you merged in develop in only one of the brancches. Please fix and I'll review.

File: gtsam/hybrid/HybridBayesNet.h [hunk 8]
```diff
@@ -59,6 +59,14 @@ class GTSAM_EXPORT HybridBayesNet : public BayesNet<AbstractConditional> {
    */
   GaussianMixture::shared_ptr atGaussian(size_t i) const;
 
+  /**
+   * @brief Set gaussian mixture at specific index.
+   *
+   * @param i The index to set the gaussian mixture at.
+   * @param gaussian The gaussian mixture shared pointer to add.
+   */
+  void setGaussian(size_t i, const GaussianMixture::shared_ptr &gaussian);
```
Comment (dellaert) [inline gtsam/hybrid/HybridBayesNet.h]: Why do we have non-const method here? Please re-evaluate.

File: gtsam/discrete/DiscreteKey.h [hunk 9]
```diff
@@ -69,6 +69,13 @@ namespace gtsam {
       push_back(key);
       return *this;
     }
+
+    /// Return the DiscreteKey vector as a set.
+    std::set<DiscreteKey> asSet() const {
```
Comment (dellaert) [inline gtsam/discrete/DiscreteKey.h]: prefer not to have changes outside hybrid. Make static function in cpp file where you need it?

File: gtsam/hybrid/IncrementalHybrid.h [hunk 10]
```diff
@@ -50,6 +50,18 @@ class IncrementalHybrid {
   void update(GaussianHybridFactorGraph graph, const Ordering& ordering,
               boost::optional<size_t> maxNrLeaves = boost::none);
 
+  /**
+   * @brief Method to prune the decision tree of discrete variables.
+   *
+   * @param maxNrLeaves The maximum number of leaves to keep.
+   * @return DecisionTreeFactor::shared_ptr 
+   */
+  DecisionTreeFactor::shared_ptr prune(size_t maxNrLeaves);
```
Comment (dellaert) [inline gtsam/hybrid/IncrementalHybrid.h]: not const. I assume this will be resolved in #44

**Review by dellaert — APPROVED** [review]
Approve but only if *both* non-const issues are resolved in #44

File: gtsam/hybrid/IncrementalHybrid.h (hunk 10 above)
Comment (varunagrawal) [reply gtsam/hybrid/IncrementalHybrid.h]: Resolved

File: gtsam/hybrid/HybridBayesNet.h (hunk 8 above)
Comment (varunagrawal) [reply gtsam/hybrid/HybridBayesNet.h]: Removed.

File: gtsam/discrete/DiscreteKey.h (hunk 9 above)
Comment (dellaert) [reply gtsam/discrete/DiscreteKey.h]: prefer not to have changes outside hybrid. Make static function in cpp file where you need it?

**Review by dellaert — CHANGES_REQUESTED** [review]

**Review by dellaert — APPROVED** [review]
Cool! Can't wait to see real results :-)

**Pull request merged**